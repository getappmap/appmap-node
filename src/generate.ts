import assert from "assert";
import type { ESTree } from "meriyah";

export function call_(
  callee: ESTree.Expression,
  ...args: ESTree.Expression[]
): ESTree.CallExpression {
  return {
    type: "CallExpression",
    callee,
    arguments: args,
  };
}

export function member(...ids: ESTree.Expression[]): ESTree.MemberExpression {
  assert(ids.length > 1);
  let result = [...ids];
  while (result.length > 1) {
    const [object, property, ...rest] = result;
    result = [
      { type: "MemberExpression", object, property, computed: property.type !== "Identifier" },
      ...rest,
    ];
  }
  assert(result[0].type === "MemberExpression");
  return result[0];
}

export type LiteralValue = ESTree.Literal["value"];

export function literal(value: LiteralValue): ESTree.Literal {
  return { type: "Literal", value };
}

export function identifier(id: string): ESTree.Identifier {
  return {
    type: "Identifier",
    name: id,
  };
}

export function ret(argument: ESTree.Expression | null = null): ESTree.ReturnStatement {
  return {
    type: "ReturnStatement",
    argument,
  };
}

export const args = identifier("arguments");
export const this_: ESTree.ThisExpression = { type: "ThisExpression" };
