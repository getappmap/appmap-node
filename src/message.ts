import console from "node:console";
import { inspect } from "node:util";

import chalk from "chalk";

export function info(message: unknown, ...args: unknown[]): void {
  console.info(prefix(message), ...args);
}

export function warn(message: unknown, ...args: unknown[]): void {
  console.warn(prefix(message), ...args);
}

function prefix(message: unknown): string {
  return chalk.magentaBright`(Î›) ` + (typeof message === "string" ? message : inspect(message));
}
