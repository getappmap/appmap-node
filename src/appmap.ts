import assert from "node:assert";
import process from "node:process";
import { fileURLToPath } from "node:url";

import type { ESTree } from "meriyah";

import type { Event as AppMapEvent } from "./event";
import { info } from "./message";
import type { Parameter } from "./parameter";
import { Event as RecorderEvent } from "./recorder";
import type { FunctionInfo } from "./registry";
import AppMapStream from "./AppMapStream";
import compactObject from "./util/compactObject";

export const stream = new AppMapStream();
process.on("exit", finish);

function resolve(event: RecorderEvent): AppMapEvent {
  const { type, id } = event;
  if (type === "call") {
    const { this_, fun, args } = event;
    const result: AppMapEvent = {
      event: type,
      id,
      thread_id: 0,
      method_id: fun.id ?? "<anonymous>",
      static: !this_,
      receiver: this_,
      parameters: resolveParameters(args, fun),
      defined_class: fun.klass,
    };
    if (fun.loc?.source?.startsWith("file://")) {
      // TODO make it relative to the root directory
      result.path = fileURLToPath(fun.loc.source);
      result.lineno = fun.loc.start.line;
    }
    return result;
  }
  assert(event.type === "return");
  return {
    event: "return",
    thread_id: 0,
    id: event.id,
    parent_id: event.parent_id,
    return_value: event.return_value,
  };
}

export function emit(event: RecorderEvent) {
  stream.emit(compactObject(resolve(event)));
}

function resolveParameters(args: Parameter[], fun: FunctionInfo): Parameter[] {
  return args.map(
    (value, index): Parameter => ({
      ...value,
      name: paramName(fun.params[index]),
    }),
  );
}

function paramName(param: ESTree.Parameter): string | undefined {
  switch (param.type) {
    case "Identifier":
      return param.name;
    case "AssignmentPattern":
      return paramName(param.left);
    // TODO: handle other parameter types
  }
}

export function finish() {
  if (stream.close()) info("Wrote %s", stream.path);
}
