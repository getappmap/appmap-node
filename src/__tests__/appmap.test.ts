/* eslint-disable @typescript-eslint/unbound-method */
import { emit, finish, stream } from "../appmap";
import { identifier } from "../generate";
import { info } from "../message";

describe(emit, () => {
  it("transforms and emits call events", () => {
    emit({
      type: "call",
      args: [
        { class: "Number", value: "42" },
        { class: "String", value: "'test'" },
      ],
      fun: {
        async: false,
        generator: false,
        id: "testfun",
        loc: {
          source: "file:///test/test.js",
          start: { line: 42, column: 1 },
          end: { line: 43, column: 0 },
        },
        params: [
          identifier("testArg"),
          { type: "AssignmentPattern", left: identifier("argWithDefault") },
        ],
      },
      id: 42,
    });

    expect(stream.emit).toBeCalledWith({
      event: "call",
      id: 42,
      thread_id: 0,
      method_id: "testfun",
      path: "/test/test.js",
      lineno: 42,
      static: true,
      parameters: [
        { class: "Number", value: "42", name: "testArg" },
        { class: "String", value: "'test'", name: "argWithDefault" },
      ],
    });
  });

  it("handles anonymous functions", () => {
    emit({
      type: "call",
      args: [],
      fun: {
        async: false,
        generator: false,
        params: [],
      },
      id: 42,
    });

    expect(stream.emit).toBeCalledWith(expect.objectContaining({ method_id: "<anonymous>" }));
  });

  it("transforms and emits return events", () => {
    emit({
      type: "return",
      id: 42,
      parent_id: 40,
    });

    expect(stream.emit).toBeCalledWith({
      event: "return",
      id: 42,
      parent_id: 40,
      thread_id: 0,
    });
  });
});

describe(finish, () => {
  it("closes the stream and prints the path if anything was written", () => {
    jest.mocked(stream.close).mockReturnValue(true);
    finish();
    expect(stream.close).toHaveBeenCalled();
    expect(jest.mocked(info).mock.lastCall).toMatchSnapshot();
  });

  it("doesn't print anything if nothing was written", () => {
    jest.mocked(stream.close).mockReturnValue(false);
    finish();
    expect(stream.close).toHaveBeenCalled();
    expect(info).not.toHaveBeenCalled();
  });
});

afterEach(jest.resetAllMocks);

Object.defineProperties(stream, {
  path: { get: () => "/test/appmap.yml", configurable: true },
  seenAny: { get: () => true, configurable: true },
});

jest.mock("../AppMapStream.ts");
jest.mock("node:fs");
jest.mock("../message");
