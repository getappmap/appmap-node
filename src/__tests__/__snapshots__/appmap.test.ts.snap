// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finish closes the stream and prints the path if anything was written 1`] = `
[
  "Wrote %s",
  "/test/appmap.yml",
]
`;
